### PR TITLE
Fixing communication set

### PR DESCRIPTION
**Pull request by hkaiser**
This fixes occasional hangs of the communication set test

### CODE REVIEW

File: libs/full/collectives/src/detail/communication_set_node.cpp [hunk 1]
```diff
@@ -105,12 +105,12 @@ namespace hpx { namespace lcos { namespace detail {
...
-            hpx::util::make_counting_iterator(num_sites), site,
-            [&](std::size_t node) {
-                return calculate_connected_node(node, arity);
-            });
+        std::size_t num_children =
+            hpx::ranges::count(hpx::util::make_counting_iterator(site + 1),
+                hpx::util::make_counting_iterator(num_sites), site,
+                [&](std::size_t node) {
+                    return calculate_connected_node(node, arity);
+                });
```
Comment (msimberg) [inline libs/full/collectives/src/detail/communication_set_node.cpp:113]: I guess this is unrelated to the hangs? (We can keep the change here, just trying to understand.)

File: libs/full/collectives/tests/unit/communication_set.cpp [hunk 2]
```diff
@@ -86,12 +86,13 @@ namespace hpx { namespace traits {
...
-        Result get(std::size_t, test::get_connected_to_zero)
+        Result get(std::size_t which, test::get_connected_to_zero)
         {
             // first forward request to parent
-            if (communicator_.connect_to_ != communicator_.site_)
+            if (communicator_.connect_to_ != communicator_.site_ &&
+                which == communicator_.site_)
             {
-                hpx::sync(test::get_connected_to_zero_action{},
+                hpx::apply(test::get_connected_to_zero_action{},
```
Comment (msimberg) [inline libs/full/collectives/tests/unit/communication_set.cpp:95]: Does this synchronize with the code after this through the mutex? I guess this is somehow the main trigger to potential deadlocks?

Comment (hkaiser) [reply libs/full/collectives/tests/unit/communication_set.cpp:95]: The test does not depend on the synchronization. And yes, this was possibly causing deadlocks.
